mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -0,0 +1,523 @@
+                };
+                Timer::sleep(30.).await;
+            }
+        }
+        .fuse(),
+    );
+
+    let ctx = swap.ctx.clone();
+    let mut status = ctx.log.status_handle();
+    let uuid_str = swap.uuid.to_string();
+    let running_swap = Arc::new(swap);
+    let weak_ref = Arc::downgrade(&running_swap);
+    let swap_ctx = SwapsContext::from_ctx(&ctx).unwrap();
+    swap_ctx.running_swaps.lock().unwrap().push(weak_ref);
+    let shutdown_rx = swap_ctx.shutdown_rx.clone();
+    let swap_for_log = running_swap.clone();
+
+    let mut swap_fut = Box::pin(
+        async move {
+            let mut events;
+            loop {
+                let res = running_swap.handle_command(command).await.expect("!handle_command");
+                events = res.1;
+                for event in events {
+                    status.status(&[&"swap", &("uuid", uuid_str.as_str())], &event.status_str());
+                    running_swap.apply_event(event);
+                }
+                match res.0 {
+                    Some(c) => {
+                        command = c;
+                    },
+                    None => {
+                        break;
+                    },
+                }
+            }
+        }
+        .fuse(),
+    );
+    let mut shutdown_fut = Box::pin(shutdown_rx.recv().fuse());
+    let do_nothing = (); // to fix https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit
+    select! {
+        _swap = swap_fut => {
+            taker_swap_watchers.remove(&uuid);
+            do_nothing
+        }, // swap finished normally
+        _shutdown = shutdown_fut => {
+            taker_swap_watchers.remove(&uuid);
+            info!("swap {} stopped!", swap_for_log.uuid)
+        },
+        _touch = touch_loop => {
+            taker_swap_watchers.remove(&uuid);
+            unreachable!("Touch loop can not stop!")
+        },
+    };
+}
+
+pub async fn process_watcher_msg(ctx: MmArc, msg: &[u8]) {
+    let msg = match decode_signed::<SwapWatcherMsg>(msg) {
+        Ok(m) => m,
+        Err(watcher_msg_err) => {
+            error!("Couldn't deserialize 'SwapWatcherMsg': {:?}", watcher_msg_err);
+            // Drop it to avoid dead_code warning
+            drop(watcher_msg_err);
+            return;
+        },
+    };
+
+    match msg.0 {
+        SwapWatcherMsg::TakerSwapWatcherMsg(watcher_data) => spawn_taker_swap_watcher(ctx, *watcher_data),
+    }
+}
+
+fn spawn_taker_swap_watcher(ctx: MmArc, watcher_data: TakerSwapWatcherData) {
+    let swap_ctx = SwapsContext::from_ctx(&ctx).unwrap();
+    let msgs = swap_ctx.swap_msgs.lock().unwrap();
+    // Return if taker or maker swap with the same uuid already exists
+    if msgs.contains_key(&watcher_data.uuid) {
+        return;
+    }
+
+    spawn(async move {
+        let swap_ctx = SwapsContext::from_ctx(&ctx).unwrap();
+        let taker_swap_watchers = match swap_ctx.taker_swap_watchers.try_lock() {
```

Comment:
I think the `taker_swap_watchers` mutex is locked for a long time in this function. Is this required? Is there a way to drop the mutex guard earlier instead of passing it to the `run_watcher` function?

Reply:
It is necessary to keep the lock until the uuid is inserted to the set, otherwise multiple can be started for the same swap. But I can release the lock manually right after inserting the uuid to the set, otherwise it will be held during the lifetime of the thread which isn't necessary.

Reply:
Maybe we can insert the uuid after this https://github.com/KomodoPlatform/atomicDEX-API/blob/63ffaa9f4546f37671240176fc56d58bb725c17b/mm2src/mm2_main/src/lp_swap/swap_watcher.rs#L479-L482 and release the lock, and remove the uuid later in case of an error. What do you think @caglaryucekaya?

Reply:
Done